### PR TITLE
Correct index on theme_user_counts

### LINKED ISSUE #5715: Add new (date, addon_id) index on theme_user_counts
Part of #5712

`theme_user_counts`:

* [x] Drop `addon_date_idx` (`addon_id`,`date`)
* [x] Add new (date, addon_id) index

### PR DESCRIPTION
**Pull request by EnTeQuAk**
Refs #5712

Fixes #5715

### CODE REVIEW

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-- Drop old index first
+DROP INDEX `addon_date_idx` ON `theme_user_counts`;
+
+-- Now create a proper
+ALTER TABLE `theme_user_counts` ADD CONSTRAINT `theme_user_counts_date_cc9034dde90789f_uniq` UNIQUE (`date`, `addon_id`)
```
Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: Generated with

```python
In [1]: from olympia.stats.models import ThemeUserCount
In [2]: from django.db.backends.mysql.schema import DatabaseSchemaEditor
In [4]: from django.db import connection
In [5]: editor = DatabaseSchemaEditor(connection)
In [8]: editor._create_unique_sql(ThemeUserCount, (ThemeUserCount._meta.get_field('date').column, ThemeUserCount._meta.get_field('addon').column))
```

… so django-migrations ready.

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+-- Drop old index first
+DROP INDEX `addon_date_idx` ON `theme_user_counts`;
+
+-- Now create a proper
+ALTER TABLE `theme_user_counts` ADD CONSTRAINT `theme_user_counts_date_cc9034dde90789f_uniq` UNIQUE (`date`, `addon_id`);
```
Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: If I drop my table locally and recreate it from `sqlall` I don't get that. I'm not sure if `sqlall` is reliable these days, though...

Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: OTOH if I create the django migration, `sqlmigrate` does result in `ALTER TABLE `theme_user_counts` ADD CONSTRAINT `theme_user_counts_date_cc9034dde90789f_uniq` UNIQUE (`date`, `addon_id`);` so it looks ok, ignore my previous comment.

Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: Yeah, `sqlall` actually throws an exception once an app uses migrations.

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+-- Drop old index first
+DROP INDEX `addon_date_idx` ON `theme_user_counts`;
```
Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: That will fail in environments where the index does not exist... I fear you need to write this in a more defensive way.

File: src/olympia/stats/models.py [hunk 4]
```diff
@@ -41,8 +41,12 @@ class StatsSearchMixin(SearchMixin):
 
 class CollectionCount(StatsSearchMixin, models.Model):
     collection = models.ForeignKey('bandwagon.Collection')
-    count = models.PositiveIntegerField()
-    date = models.DateField()
+
+    # index name in database: `count`
```
Comment (diox) [inline src/olympia/stats/models.py]: Everytime you mention `in database` it should read `in dev/stage/prod database` or something, just to remove any confusion.

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql (hunk 3 above)
Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: Fixed in 14a48c27d4

File: src/olympia/stats/models.py (hunk 4 above)
Comment (EnTeQuAk) [inline src/olympia/stats/models.py]: Should be fixed in aec83ec4aa

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql [hunk 5]
```diff
@@ -0,0 +1,21 @@
+-- Drop old index first, if it exists.
+
+-- Set the index_name or NULL if it doesn't exist
+SET @KEY_ADDON_DATE_IDX := (
+    SELECT INDEX_NAME FROM information_schema.STATISTICS WHERE
+        TABLE_SCHEMA = (SELECT DATABASE()) AND
+        TABLE_NAME = 'theme_user_counts' AND
+        INDEX_NAME = 'addon_date_idx'
+    LIMIT 1);
```
Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: This is needed because we have 2 columns that match on the `INDEX_NAME`.

Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: uh ?

Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: the index is a composite key on (addon_id, date) - so it matches two columns in that query. That's why we need the limit 1.

Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: aah ok

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+SET @QUERY_DROP_KEY_ADDON_DATE_IDX = IF(
+    @KEY_ADDON_DATE_IDX IS NOT NULL,
+    CONCAT('ALTER TABLE validation_job DROP KEY ', @KEY_ADDON_DATE_IDX, ';'),
```
Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: Wrong copy/paste

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql (hunk 5 above)
Comment (diox) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: Might want to clarify that in a comment in the migration itself

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql (hunk 6 above)
Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: ah, sorry about that, fixed in ec51619063812eb7ad3b943b61ea20053b38750b

File: src/olympia/migrations/952-correct-index-theme-user-counts.sql (hunk 5 above)
Comment (EnTeQuAk) [inline src/olympia/migrations/952-correct-index-theme-user-counts.sql]: added in ec51619063812eb7ad3b943b61ea20053b38750b

**Pull request merged**